Xext/panoramiX.c

```diff
@@ -1064,28 +1064,31 @@ ProcXineramaQueryScreens(ClientPtr client)
         swapl(&rep.length);
         swapl(&rep.number);
     }
-    WriteToClient(client, sizeof(xXineramaQueryScreensReply), &rep);
+
+    xXineramaScreenInfo *scratch = calloc(number, sizeof(xXineramaScreenInfo));
+    if (number && !scratch)
+        return BadAlloc;
 
     if (!noPanoramiXExtension) {
-        xXineramaScreenInfo scratch;
         int i;
-
         FOR_NSCREENS_BACKWARD(i) {
-            scratch.x_org = screenInfo.screens[i]->x;
-            scratch.y_org = screenInfo.screens[i]->y;
-            scratch.width = screenInfo.screens[i]->width;
-            scratch.height = screenInfo.screens[i]->height;
+            scratch[i].x_org = screenInfo.screens[i]->x;
+            scratch[i].y_org = screenInfo.screens[i]->y;
+            scratch[i].width = screenInfo.screens[i]->width;
+            scratch[i].height = screenInfo.screens[i]->height;
 
             if (client->swapped) {
-                swaps(&scratch.x_org);
-                swaps(&scratch.y_org);
-                swaps(&scratch.width);
-                swaps(&scratch.height);
+                swaps(&scratch[i].x_org);
+                swaps(&scratch[i].y_org);
+                swaps(&scratch[i].width);
+                swaps(&scratch[i].height);
             }
-            WriteToClient(client, sz_XineramaScreenInfo, &scratch);
         }
     }
 
+    WriteToClient(client, sizeof(xXineramaQueryScreensReply), &rep);
+    WriteToClient(client, number * sizeof(xXineramaQueryScreensReply), scratch);
```

Comment:
Since the result of the above `calloc()` with zero is non-portable/implementation-defined, not exactly nice to potentially pass null here, even if with zero length, while the call shouldn't even be made since there's nothing to write.